Add script and docs for preparing releases (#720)
fitbenchmarking/fitbenchmarking: docs/source/contributors/workflow.rst

If you encounter problems with cherry picking into release branch please
don't hesitate to speak to an experienced member of the FitBenchmarking team.

================

> Create a release

Reviewer: ```suggestion
Creating a release
```

================
In order to create a new release for FitBenchmarking, there are a few manual steps.
These have been streamlined as much as possible.

> First checkout the branch to create the release from.

Reviewer: ```suggestion
First checkout the branch to create the release from.  Releases should only be made from a `release-x-x` branch, not a development branch or master.
```

Reviewer: I can't think of a reason why we'd not do this -- can you?  We should probably enforce this in the script.

Author: I agree, might as well check.

From the root of the repo run the "ci/prep_and_tag_release.sh" script with the new version number.
The version number will be rejected if it is not of the expected form.
We expect a "v" followed by the major, minor, and patch numbers,
and an optional numbered label to mark the type of release.

> Possible lables are:

Reviewer: ```suggestion
Possible labels are:
```

 - -beta (release for testing)
 - -rc (release candidate)
 - -dev (release of a development version)

> 

Reviewer: I know we have these three already, but do we need them?  What's the difference between `dev` and `beta`, for example?

Author: I was just listing what we already had... Maybe we'd want a `dev` if someone requested a bleeding edge feature and we weren't near a `beta`? We can probably remove this until we need it.

>  - -dev (release of a development version)

Author: ```suggestion
```

This script will create a new commit with the docs and testing links updated, tag it,
and revert the change in a second commit so that the links point back to the latest versions.

These commits will need to be pushed to github.

Finally, you will need to create a release on github.
This can be done by navigating to the releases page, selecting new release
and typing in the tag that was given to the release
(it should tell you the tag exists at this point!).

> E.g For a first beta version of release 0.1.0. One would run:

Reviewer: ```suggestion
For example, For a first beta version of release 0.1.0, one would run:
```
